src/main/java/bio/terra/cli/service/GoogleBigQuery.java

```diff
@@ -0,0 +1,130 @@
+  }
+
+  private GoogleBigQuery(GoogleCredentials credentials) {
+    bigQuery = CrlUtils.createBigQueryCow(credentials);
+  }
+
+  public Optional<Dataset> getDataset(String projectId, String datasetId) {
+    try {
+      Dataset dataset =
+          callWithRetries(
+              () -> bigQuery.datasets().get(projectId, datasetId).execute(),
+              "Error looking up dataset.");
+      return Optional.ofNullable(dataset);
+    } catch (Exception ex) {
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Returns the number of tables in the dataset, or null if there was an error looking it up. This
+   * behavior is useful for display purposes.
+   */
+  public Integer getNumTables(String projectId, String datasetId) {
+    try {
+      TableList tables =
+          callWithRetries(
+              () -> bigQuery.tables().list(projectId, datasetId).execute(),
+              "Error looking up dataset tables.");
+      return tables.getTotalItems();
+    } catch (Exception ex) {
+      return null;
+    }
+  }
+
+  /**
+   * Execute a function that includes hitting BQ endpoints. Retry if the function throws an {@link
+   * #isRetryable} exception. If an exception is thrown by the BQ client or the retries, make sure
+   * the HTTP status code and error message are logged.
+   *
+   * @param makeRequest function with a return value
+   * @param errorMsg error message for the the {@link SystemException} that wraps any exceptions
+   *     thrown by the BQ client or the retries
+   */
+  private <T> T callWithRetries(
+      HttpUtils.SupplierWithCheckedException<T, IOException> makeRequest, String errorMsg) {
+    return handleClientExceptions(
+        () ->
+            HttpUtils.callWithRetries(
+                makeRequest,
+                GoogleBigQuery::isRetryable,
+                BQ_MAXIMUM_RETRIES,
+                HttpUtils.DEFAULT_DURATION_SLEEP_FOR_RETRY),
+        errorMsg);
+  }
+
+  /**
+   * Execute a function that includes hitting BQ endpoints. If an exception is thrown by the BQ
+   * client or the retries, make sure the HTTP status code and error message are logged.
+   *
+   * @param makeRequest function with a return value
+   * @param errorMsg error message for the the {@link SystemException} that wraps any exceptions
+   *     thrown by the BQ client or the retries
+   */
+  private <T> T handleClientExceptions(
+      HttpUtils.SupplierWithCheckedException<T, IOException> makeRequest, String errorMsg) {
+    try {
+      return makeRequest.makeRequest();
+    } catch (IOException | InterruptedException ex) {
+      // wrap the BQ exception and re-throw it
+      throw new SystemException(errorMsg, ex);
+    }
+  }
+
+  /**
+   * Utility method that checks if an exception thrown by the BQ client is retryable.
+   *
+   * @param ex exception to test
+   * @return true if the exception is retryable
+   */
+  static boolean isRetryable(Exception ex) {
```

Comment:
Does this need to handle request timeouts? I'm not exactly sure how those are thrown when you're calling CRL, or if that's already handled somewhere else

Reply:
Do you mean a `SocketTimeoutException`? I'm not sure either, but since those CRL methods declare an `IOException` in their signature, makes sense to me to retry those here.